### PR TITLE
affine kvm-pit kernel thread to vcpu-0 cpumask

### PR DESCRIPTION
**Pull request by matosatti**
**What this PR does / why we need it**:
Certain use-cases rely on injection of PIT interrupts to the guest in a timely fashion. Today, the kvm-pit kernel thread which is responsible for emulating PIT interrupts runs on housekeeping cores, sharing time with other kubernetes threads (such as kubelet). This has been found to be a problem for a particular workload requiring timely PIT interrupts (see trace below). The fix is to affine kvm-pit kernel thread to the same cpu as vcpu-0, when dedicated cpu placement is enabled in the guest configuration.

          <idle>-0     [000] d.h. 102470.847586: hrtimer_expire_entry: hrtimer=00000000e3dea339 function=pit_timer_fn [kvm] now=102463903132805
          <idle>-0     [000] d.h. 102470.847587: sched_waking: comm=kvm-pit/67 pid=1212317 prio=120 target_cpu=001
          <idle>-0     [001] d... 102470.847590: sched_switch: prev_comm=swapper/1 prev_pid=0 prev_prio=120 prev_state=S ==> next_comm=kvm-pit/67 next_pid=1212317 next_prio=120
      kvm-pit/67-1212317 [001] d... 102470.847594: sched_switch: prev_comm=kvm-pit/67 prev_pid=1212317 prev_prio=120 prev_state=D ==> next_comm=swapper/1 next_pid=0 next_prio=120
       CPU 0/KVM-1212312 [006] d... 102470.847596: kvm_entry: vcpu 0, rip 0x35fd
       CPU 0/KVM-1212312 [006] d... 102470.847597: kvm_exit: vcpu 0 reason EOI_INDUCED rip 0x1bab info1 0x0000000000000022 info2 0x0000000000000000 intr_info 0x00000000 error_code 0x00000000
       CPU 0/KVM-1212312 [006] d... 102470.847598: kvm_entry: vcpu 0, rip 0x1bab
...
       CPU 0/KVM-1212312 [006] d... 102470.847622: kvm_exit: vcpu 0 reason HLT rip 0x35fc info1 0x0000000000000000 info2 0x0000000000000000 intr_info 0x00000000 error_code 0x00000000
           <...>-64244 [000] d.h. 102470.857590: hrtimer_expire_entry: hrtimer=00000000e3dea339 function=pit_timer_fn [kvm] now=102463913135899
           <...>-64244 [000] d.h. 102470.857599: sched_waking: comm=kvm-pit/67 pid=1212317 prio=120 target_cpu=001
  kube-apiserver-64649 [004] d.h. 102470.860004: hrtimer_expire_entry: hrtimer=00000000f5450c52 function=pit_timer_fn [kvm] now=102463915550678
           <...>-64245 [003] d.h. 102470.863029: hrtimer_expire_entry: hrtimer=000000004736fe96 function=pit_timer_fn [kvm] now=102463918575518
           <...>-64244 [000] d.h. 102470.867587: hrtimer_expire_entry: hrtimer=00000000e3dea339 function=pit_timer_fn [kvm] now=102463923133350
           <...>-64244 [000] d.h. 102470.869780: hrtimer_expire_entry: hrtimer=000000005deb41ca function=pit_timer_fn [kvm] now=102463925326017
           <...>-64244 [000] d... 102470.877465: sched_switch: prev_comm=kube-apiserver prev_pid=64244 prev_prio=120 prev_state=S ==> next_comm=kvm-pit/67 next_pid=1212317 next_prio=120
      kvm-pit/67-1212317 [000] d... 102470.877483: sched_switch: prev_comm=kvm-pit/67 prev_pid=1212317 prev_prio=120 prev_state=D ==> next_comm=cluster-policy- next_pid=60163 next_prio=120
       CPU 0/KVM-1212312 [006] d... 102470.877486: kvm_entry: vcpu 0, rip 0x35fd
       CPU 0/KVM-1212312 [006] d... 102470.877490: kvm_exit: vcpu 0 reason EOI_INDUCED rip 0x1bab info1 0x0000000000000022 info2 0x0000000000000000 intr_info 0x00000000 error_code 0x00000000
       CPU 0/KVM-1212312 [006] d... 102470.877491: sched_waking: comm=kvm-pit/67 pid=1212317 prio=120 target_cpu=000
       CPU 0/KVM-1212312 [006] d... 102470.877494: kvm_entry: vcpu 0, rip 0x1bab

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note NONE

```

### CODE REVIEW

File: pkg/virt-handler/vm.go [hunk 1]
```diff
@@ -2698,6 +2700,110 @@ func (d *VirtualMachineController) vmUpdateHelperMigrationTarget(origVMI *v1.Vir
 	return nil
 }
 
+func get_nspid(vmpid int, vmi *v1.VirtualMachineInstance) (int, error) {
+	fpath := "/proc/" + strconv.Itoa(vmpid) + "/status"
```
Comment (vladikr) [inline pkg/virt-handler/vm.go:2704]: could you please use filepath.Join ?

File: pkg/virt-handler/vm.go [hunk 2]
```diff
@@ -2698,6 +2700,110 @@ func (d *VirtualMachineController) vmUpdateHelperMigrationTarget(origVMI *v1.Vir
 	return nil
 }
 
+func get_nspid(vmpid int, vmi *v1.VirtualMachineInstance) (int, error) {
```
Comment (vladikr) [inline pkg/virt-handler/vm.go:2703]: Can you please add more details as to what does this function gets?

File: pkg/virt-handler/vm.go [hunk 3]
```diff
@@ -2921,6 +3027,12 @@ func (d *VirtualMachineController) vmUpdateHelperDefault(origVMI *v1.VirtualMach
 			return err
 		}
 	}
+	if vmi.IsCPUDedicated() && !vmi.IsRunning() && !vmi.IsFinal() {
+		log.Log.V(3).Object(vmi).Info("Affining PIT thread")
+		if err := d.affinePitThread(vmi); err != nil {
+			return err
+		}
+	}
```
Comment (vladikr) [inline pkg/virt-handler/vm.go:3035]: :+1:

File: pkg/virt-handler/vm.go [hunk 4]
```diff
@@ -2698,6 +2700,110 @@ func (d *VirtualMachineController) vmUpdateHelperMigrationTarget(origVMI *v1.Vir
...
+		return err
+	}
+	if qemupid == -1 {
+		return nil
+	}
+
+	pitpid := kvmPitPid(qemupid, vmi)
+	if pitpid == -1 {
+		return nil
+	}
```
[the remark this answers is not in the document]
Comment (matosatti) [reply pkg/virt-handler/vm.go]: @vladikr hopefully all of your comments have been addressed.

**Comment by vladikr** [comment]
Thanks @matosatti ! 
/approve

@acardace could you please take a look?

File: pkg/virt-handler/isolation/isolation.go [hunk 5]
```diff
@@ -59,6 +64,10 @@ type IsolationResult interface {
 	Mounts(mount.FilterFunc) ([]*mount.Info, error)
 	// returns the QEMU process
 	GetQEMUProcess() (ps.Process, error)
+	// returns the KVM PIT pid
+	KvmPitPid(vmpid int, vmi *v1.VirtualMachineInstance) (int, error)
```
Comment (xpivarc) [inline pkg/virt-handler/isolation/isolation.go:68]: `vmpid` seems to be not required as argument as you can just call `GetQEMUProcess` inside `KvmPitPid`

File: pkg/virt-handler/isolation/isolation.go [hunk 6]
```diff
@@ -160,6 +169,71 @@ func (r *RealIsolationResult) GetQEMUProcess() (ps.Process, error) {
...
+
+func (r *RealIsolationResult) KvmPitPid(vmpid int, vmi *v1.VirtualMachineInstance) (int, error) {
+	processes, _ := ps.Processes()
+	nspid, err := GetNspid(vmpid, vmi)
+	if err != nil || nspid == -1 {
+		return -1, err
+	}
+	pitstr := "kvm-pit/" + strconv.Itoa(nspid)
+
+	idx := 0
```
Comment (xpivarc) [inline pkg/virt-handler/isolation/isolation.go:212]: This does seems to be not used.

**Review by xpivarc — CHANGES_REQUESTED** [review]
Few nits inline

File: pkg/virt-handler/isolation/isolation.go [hunk 7]
```diff
@@ -160,6 +169,71 @@ func (r *RealIsolationResult) GetQEMUProcess() (ps.Process, error) {
...
+	file, err := os.Open(fpath)
+	if err != nil {
+		return 0, err
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := scanner.Text()
+		fmt.Println(line)
```
[the remark this answers is not in the document]
Comment (matosatti) [reply pkg/virt-handler/isolation/isolation.go]: Fixed

File: pkg/virt-handler/isolation/isolation.go [hunk 8]
```diff
@@ -160,6 +169,71 @@ func (r *RealIsolationResult) GetQEMUProcess() (ps.Process, error) {
...
+	for _, process := range processes {
+		idx = idx + 1
+		if process.Executable() == pitstr {
+			return process.Pid(), nil
+		}
+	}
+	return -1, nil
+}
+
+func (r *RealIsolationResult) GetQEMUKVMProcessPid() (int, error) {
```
[the remark this answers is not in the document]
Comment (matosatti) [reply pkg/virt-handler/isolation/isolation.go]: Fixed

**Comment by matosatti** [comment]
> Few nits inline

Luboslav,

Hopefully addressed all of your comments. Let me know if it looks OK now.

File: pkg/virt-handler/isolation/isolation.go [hunk 9]
```diff
@@ -160,6 +167,52 @@ func (r *RealIsolationResult) GetQEMUProcess() (ps.Process, error) {
 	return qemuProcess, nil
 }
 
+// Returns the pid of "vmpid" as seen from the first pid namespace the task
+// belongs to.
+func GetNspid(vmpid int, vmi *v1.VirtualMachineInstance) (int, error) {
```
Comment (xpivarc) [inline pkg/virt-handler/isolation/isolation.go:172]: Argument `vmi` is unused.

File: pkg/virt-handler/isolation/isolation.go [hunk 10]
```diff
@@ -160,6 +167,52 @@ func (r *RealIsolationResult) GetQEMUProcess() (ps.Process, error) {
...
+			continue
+		}
+		val, err := strconv.Atoi(s[2])
+		return val, err
+	}
+
+	return 0, nil
+}
+
+func (r *RealIsolationResult) KvmPitPid(vmpid int, vmi *v1.VirtualMachineInstance) (int, error) {
```
Comment (xpivarc) [inline pkg/virt-handler/isolation/isolation.go:200]: You don't need `vmi` argument as it is unused in `GetNspid`.

File: pkg/virt-handler/isolation/isolation.go (hunk 5 above)
Comment (xpivarc) [reply pkg/virt-handler/isolation/isolation.go:68]: What I meant is that the struct implementing this interface also needs to have `GetQEMUProcess` method. Meaning you can just call it within `KvmPitPId`

```
func (r *RealIsolationResult) KvmPitPid() (int, error) {
        vmpid, err := r.GetQEMUProcess()
        if err != nil {
           return err
         }
	processes, _ := ps.Processes()
	nspid, err := GetNspid(vmpid)
	if err != nil || nspid == -1 {
		return -1, err
	}
	pitstr := "kvm-pit/" + strconv.Itoa(nspid)

	for _, process := range processes {
		if process.Executable() == pitstr {
			return process.Pid(), nil
		}
	}
	return -1, nil
```

**Review by xpivarc — COMMENTED** [review]
You need to run `make generate` locally.

**Comment by matosatti** [comment]
> You need to run `make generate` locally.

@xpivarc your comments should have been addressed, thanks!

File: pkg/virt-handler/isolation/isolation.go [hunk 11]
```diff
@@ -160,6 +165,56 @@ func (r *RealIsolationResult) GetQEMUProcess() (ps.Process, error) {
...
+		}
+		s := strings.Fields(line)
+		if len(s) < 2 {
+			continue
+		}
+		val, err := strconv.Atoi(s[2])
+		return val, err
+	}
+
+	return 0, nil
```
[the remark this answers is not in the document]
Comment (matosatti) [reply pkg/virt-handler/isolation/isolation.go]: If a process does not have "NSpid" field in its status file, yes it can.

**Review by xpivarc — APPROVED** [review]

**Pull request merged**